File: clients/javascript/game/cell.js
const Vector = require('victor');

const CellAction = require('./cell-action.js');

module.exports = class Cell {
  constructor(id, mass, radius, position, target) {
    this.id = id;
    this.mass = mass;
    this.radius = radius;
    this.position = position;
    this.target = target;

    this.actions_ = new CellAction(id);
  }

  static parse(payload) {
    return new Cell(
      payload.id,
      payload.mass,
      payload.radius,
      Vector.fromObject(payload.position),
      Vector.fromObject(payload.target)
    );
  }

  move(target) {
    this.actions_.target = target;
  }

  burst() {
    this.actions_.burst = true;
  }

  split() {
    this.actions_.split = true;
  }

  trade(t) {
    this.actions_.trade = t;
  }

  actions() {
    const actions = this.actions_.export();

    if(actions && !actions.target) {
      actions.target = this.target;
    }

    return actions;
  }
};

File: clients/java/src/main/java/io/aigar/game/models/Cell.java
package io.aigar.game.models;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.mini2Dx.gdx.math.Vector2;

import java.util.Optional;

public class Cell {
    private int id;
    private int mass;
    private int radius;
    private Vector2 position;
    private Vector2 target;
    private boolean burst;
    @JsonIgnore
    private CellActions cellActions;

    public int getId() {
        return id;
    }

    public int getMass() {
        return mass;
    }

    public int getRadius() {
        return radius;
    }

    public Vector2 getPosition() {
        return position;
    }

    public Vector2 getTarget() {
        return target;
    }

    public boolean isBurst() {
        return burst;
    }

    @JsonIgnore
    public void move(Vector2 target) {
        cellActions.setTarget(target);
    }

    @JsonIgnore
    public void split() {
        cellActions.setSplit(true);
    }

    @JsonIgnore
    public void burst() {
        cellActions.setBurst(true);
    }

    @JsonIgnore
    public void trade(int quantity) {
        cellActions.setTrade(quantity);
    }

    @JsonIgnore
    public void initCellActions() {
        cellActions = new CellActions();
    }

    @JsonIgnore
    public Optional<CellActions> getActions() {
        if(!cellActions.getChanged())
            return Optional.empty();

        if(cellActions.getTarget() == null)
            cellActions.setTarget(target);

        return Optional.of(cellActions.withCellId(id));
    }
}
